src/types/mappingproxy.rs

```diff
@@ -0,0 +1,563 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use super::PyMapping;
+use crate::err::PyResult;
+use crate::types::dict::PyDictItem;
+#[cfg(all(test, not(PyPy)))]
+use crate::types::dict::{PyDictItems, PyDictKeys, PyDictValues};
+use crate::types::{IntoPyDict, PyAny, PyDict, PyIterator, PyList, PySequence};
+use crate::PyObject;
+#[cfg(not(PyPy))]
+use crate::{ffi, AsPyPointer, PyErr, PyTryFrom, Python, ToPyObject};
+use std::os::raw::c_int;
+
+#[inline]
+#[allow(non_snake_case)]
+unsafe fn PyDictProxy_Check(object: *mut crate::ffi::PyObject) -> c_int {
+    ffi::PyObject_TypeCheck(object, ffi::addr_of_mut_shim!(ffi::PyDictProxy_Type))
+}
+
+/// Represents a Python `mappingproxy`.
+#[repr(transparent)]
+pub struct PyMappingProxy(PyAny);
+
+pyobject_native_type!(
+    PyMappingProxy,
+    ffi::PyDictProxyObject,
+    ffi::PyDictProxy_Type,
+    #checkfunction=PyDictProxy_Check
+);
```

Comment:
Because we don't have a definition for `PyDictProxyObject` (as above), we should instead do:

```suggestion
pyobject_native_type_core!(
    PyMappingProxy,
    ffi::PyDictProxy_Type,
    #checkfunction=PyDictProxy_Check
);
```

This is almost equivalent, essentially just means that we can't use `#[pyclass(extends=PyMappingProxy)]` because we don't know the layout of the struct in memory.